src/main/java/servlets/UploadReceiptServlet.java

```diff
@@ -173,4 +164,65 @@ private static boolean isValidFilename(String filename) {
   private String getBlobServingUrl(BlobKey blobKey) {
     return "/serve-image?blob-key=" + blobKey.getKeyString();
   }
+
+  /**
+   * Extracts the raw text from the image with the Cloud Vision API. Returns a receipt
+   * entity populated with the extracted fields.
+   */
+  private Entity analyzeReceiptImage(String imageUrl, HttpServletRequest request)
+      throws ReceiptAnalysisException {
+    String absoluteUrl = getBaseUrl(request) + imageUrl;
+    AnalysisResults results = null;
+
+    try {
+      results = ReceiptAnalysis.serveImageText(absoluteUrl);
+    } catch (IOException e) {
+      throw new ReceiptAnalysisException("Receipt analysis failed.", e);
+    }
+
+    // TODO: Replace hard-coded values using receipt analysis with Cloud Vision.
+    double price = 5.89;
+    String store = "McDonald's";
+
+    // Create an entity with a kind of Receipt.
+    Entity receipt = new Entity("Receipt");
+    receipt.setProperty("price", price);
+    receipt.setProperty("store", store);
+    receipt.setUnindexedProperty("rawText", results.getRawText());
+
+    return receipt;
+  }
+
+  /**
+   * Get the base URL of the application for either the dev server or deployment.
+   */
+  private String getBaseUrl(HttpServletRequest request) {
+    String baseUrl = request.getScheme() + "://" + request.getServerName() + ":"
+        + request.getServerPort() + request.getContextPath();
+
+    // URL for dev server.
+    if (baseUrl.equals("http://0.0.0.0:80")) {
+      return "https://8080-9333ac09-1dd5-4f7f-8a7a-34f16c364c6b.us-east1.cloudshell.dev";
```

Comment:
Is it possible to avoid hardcoding this URL? It'll be different for Emily and me.

Reply:
For this commit, accessing receipt images on the back-end from the URL was not working on the dev server due to authentication issues with Cloud Shell. 

I looked into a few alternative solutions like getting resources through the servlet context and using a request dispatcher, and considering that user authentication checks will be added to the servlets for the MVP, I think that the best option is fetching the bytes with the blob key. The latest commit uses the imageUrl in deployment and the blob key for the dev server to access the receipt image from the shared ReceiptAnalysis class.